More readable code and major fixes for form reset

paavininanda opened the pull request:
<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix problem stated in  https://github.com/servo/servo/pull/19963 by @jonleighton 

<!-- Either: -->
- [x] There are tests for these changes
<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/20069)
<!-- Reviewable:end -->

paavininanda committed:
More readable code and fixes

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/htmltextareaelement.rs, components/script/dom/htmlformelement.rs, components/script/dom/htmlinputelement.rs, components/script/textinput.rs
 * @fitzgen: components/script/dom/htmltextareaelement.rs, components/script/dom/htmlformelement.rs, components/script/dom/htmlinputelement.rs, components/script/textinput.rs
 * @kichjang: components/script/dom/htmltextareaelement.rs, components/script/dom/htmlformelement.rs, components/script/dom/htmlinputelement.rs, components/script/textinput.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits include an empty title element (`<title></title>`). Consider adding appropriate metadata.

servo-wpt-sync commented:
Opened new PR for upstreamable changes.

Completed upstream sync of web-platform-test changes at https://github.com/jdm/web-platform-tests/pull/29.

paavininanda commented:
r? @jdm

CYBAI commented on components/script/textinput.rs:
```diff
@@ -850,7 +857,7 @@ impl<T: ClipboardProvider> TextInput<T> {
             vec!(content)
         };
-        if update_text_cursor {
+        if update_text_cursor == ChangeEditPoint::Change {
```
I'm wondering if we can use `if let` here? Then maybe we won't need to `clone()` for the `update_text_cursor` argument in `htmltextareaelement.rs`.

```rust
pub fn set_content(&mut self, content: DOMString, update_text_cursor: &ChangeEditPoint) {
    ....
    if let ChangeEditPoint::Change = *update_text_cursor {
        self.edit_point.line = min(self.edit_point.line, self.lines.len() - 1);
        self.edit_point.index = min(self.edit_point.index, self.current_line_length());
   }
    ....
}
```

If the way in above example is fine, I think it would be great to also use the same way for `clear_selection_to_limit`! :)

CYBAI commented on components/script/textinput.rs:
```diff
@@ -625,11 +631,12 @@ impl<T: ClipboardProvider> TextInput<T> {
 
-    pub fn adjust_horizontal_to_limit(&mut self, direction: Direction, select: Selection, update_text_cursor: bool) {
+    pub fn adjust_horizontal_to_limit(&mut self, direction: Direction,
+        select: Selection, update_text_cursor: ChangeEditPoint) {
```
nit: I think it would be great to have the indentation like following

```rust
pub fn adjust_horizontal_to_limit(
    &mut self,
    direction: Direction,
    select: Selection,
    update_text_cursor: ChangeEditPoint
) {
```

CYBAI commented:
Errors in Travis

```
$ ./mach test-tidy --no-progress --all
Checking the config file...
Checking the wpt manifest file...
Checking directories for correct file extensions...
Running the WPT lint on ./tests/wpt/web-platform-tests/...
{"path": "html/semantics/forms/resetting-a-form/reset-form-after-updating-input-textarea.html", "message": "Tabs used for indentation", "lineno": 11, "rule": "INDENT TABS"}
{"path": "html/semantics/forms/resetting-a-form/reset-form-after-updating-input-textarea.html", "message": "Tabs used for indentation", "lineno": 12, "rule": "INDENT TABS"}
{"path": "html/semantics/forms/resetting-a-form/reset-form-after-updating-input-textarea.html", "message": "Tabs used for indentation", "lineno": 13, "rule": "INDENT TABS"}
{"path": "html/semantics/forms/resetting-a-form/reset-form-after-updating-input-textarea.html", "message": "Tabs used for indentation", "lineno": 14, "rule": "INDENT TABS"}
{"path": "html/semantics/forms/resetting-a-form/reset-form-after-updating-input-textarea.html", "message": "Tabs used for indentation", "lineno": 15, "rule": "INDENT TABS"}
{"path": "html/semantics/forms/resetting-a-form/reset-form-after-updating-input-textarea.html", "message": "Tabs used for indentation", "lineno": 16, "rule": "INDENT TABS"}
{"path": "html/semantics/forms/resetting-a-form/reset-form-after-updating-input-textarea.html", "message": "Tabs used for indentation", "lineno": 17, "rule": "INDENT TABS"}
{"path": "html/semantics/forms/resetting-a-form/reset-form-after-updating-input-textarea.html", "message": "Tabs used for indentation", "lineno": 19, "rule": "INDENT TABS"}
{"path": "html/semantics/forms/resetting-a-form/reset-form-after-updating-input-textarea.html", "message": "Tabs used for indentation", "lineno": 20, "rule": "INDENT TABS"}
{"path": "html/semantics/forms/resetting-a-form/reset-form-after-updating-input-textarea.html", "message": "Tabs used for indentation", "lineno": 21, "rule": "INDENT TABS"}
{"path": "html/semantics/forms/resetting-a-form/reset-form-after-updating-input-textarea.html", "message": "Tabs used for indentation", "lineno": 22, "rule": "INDENT TABS"}
{"path": "html/semantics/forms/resetting-a-form/reset-form-after-updating-input-textarea.html", "message": "Tabs used for indentation", "lineno": 23, "rule": "INDENT TABS"}
{"path": "html/semantics/forms/resetting-a-form/reset-form-after-updating-input-textarea.html", "message": "Tabs used for indentation", "lineno": 24, "rule": "INDENT TABS"}
{"path": "html/semantics/forms/resetting-a-form/reset-form-after-updating-input-textarea.html", "message": "Tabs used for indentation", "lineno": 26, "rule": "INDENT TABS"}
{"path": "html/semantics/forms/resetting-a-form/reset-form-after-updating-input-textarea.html", "message": "Tabs used for indentation", "lineno": 28, "rule": "INDENT TABS"}
{"path": "html/semantics/forms/resetting-a-form/reset-form-after-updating-input-textarea.html", "message": "Tabs used for indentation", "lineno": 29, "rule": "INDENT TABS"}
{"path": "html/semantics/forms/resetting-a-form/reset-form-after-updating-input-textarea.html", "message": "Tabs used for indentation", "lineno": 30, "rule": "INDENT TABS"}
{"path": "html/semantics/forms/resetting-a-form/reset-form-after-updating-input-textarea.html", "message": "Tabs used for indentation", "lineno": 31, "rule": "INDENT TABS"}
{"path": "html/semantics/forms/resetting-a-form/reset-form-after-updating-input-textarea.html", "message": "Tabs used for indentation", "lineno": 32, "rule": "INDENT TABS"}
{"path": "html/semantics/forms/resetting-a-form/reset-form-after-updating-input-textarea.html", "message": "Tabs used for indentation", "lineno": 33, "rule": "INDENT TABS"}
WPT Lint Tool:: lint error(s) in Web Platform Tests: exit status 20
Running the WPT lint on ./tests/wpt/mozilla/tests/...
Checking files for tidiness...
Running the dependency licensing lint...

The command "./mach test-tidy --no-progress --all" exited with 1.
```

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at jdm/web-platform-tests#29.

jonleighton commented:
Hi @paavininanda, did you explore whether it's possible to re-implement the `reset` method to avoid any of these parameters? (I.e. what I was getting at in https://github.com/servo/servo/pull/19963#issuecomment-365632073 ?)

If that works, I think it would be best to make 2 commits as follows:

* Commit 1 to [revert](https://www.atlassian.com/git/tutorials/undoing-changes/git-revert) your previous commit which introduced the boolean parameters
* Commit 2 to re-implement `reset` on `HTMLInputElement` and `HTMLTextAreaElement` so that instead of calling `self.SetValue(...)`, they set the "raw value" by calling `self.textinput.borrow_mut().set_content(...)`. This commit would also include the WPT test code you've written.

paavininanda committed:
Removed unnecessary fn parameters

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at jdm/web-platform-tests#29.

paavininanda commented:
@jonleighton I have removed unnecessary function parameters. I also changed the reset function. All the tests are passing. I hope this is what you were looking for.

jonleighton commented:
@paavininanda Thanks, this is moving in the right direction!

However, I think it still needs a couple more things:

1. I don't think we now need the extra parameter anywhere. The behaviour in all cases is now the same. So you can completely remove the `ChangeEditPoint` enum, and the `update_text_cursor` parameter.
2. As I said above I think it would be better to make this change in two commits: one to revert [your previous commit which introduced the boolean parameters](https://github.com/servo/servo/commit/b2c1f89b932a72f9e0110c17adde33647e84c902), and then one to add the new fix (changing the `reset` implementation) and the test. I think this is better because it will make the history of the changes clearer: we are basically abandoning one approach and now trying a completely new approach. This will help other developers when they come to review the git history to understand why certain changes have been made.

As for how you would achieve number 2, I would do it like this:

1. Continue making changes on this branch until you're happy with them
2. [Squash](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html) all the commits together into one
3. Switch to the master branch: `git checkout master`
4. Make a new branch off master: `git checkout -b [your-branch-name]`
5. Revert the previous commit: `git revert b2c1f89b932a72f9e0110c17adde33647e84c902`
6. Find out the SHA1 of the commit from this branch which contains the changes you want to apply on top of the revert: `git log --oneline -n 1 Issue19963n`
7. [Cherry pick](http://think-like-a-git.net/sections/rebase-from-the-ground-up/cherry-picking-explained.html) that commit onto the new branch: `git cherry-pick [the-commit-SHA1]`
8. Now make sure that the commit is *only* making changes to the tests and the reset method (review the changes with `git show`, modify the commit if needed with `git commit -a --amend`)

jonleighton commented:
Regarding the test, I think that we should probably only be testing the initial value of selectionStart/End (i.e. the original thing we were trying to fix) since there are [steps being taken upstream](https://github.com/whatwg/html/pull/3472) to specify/test the other edge cases (like what happens after reset) and this might just duplicate effort. (@jdm may disagree, I'm not sure..!)

paavininanda commented on components/script/textinput.rs:
```diff
@@ -171,7 +177,7 @@ impl<T: ClipboardProvider> TextInput<T> {
             selection_direction: selection_direction,
         };
-        i.set_content(initial, false);
+        i.set_content(initial, &ChangeEditPoint::NoChange);
```
@jonleighton I think we need it because of this. set_content here needs no change. All other set_content calls need change

jdm replied:
The callers of TextInput::new in htmltextarea.rs and htmlinputelement.rs provide an empty string, so this does not actually help us.

jdm commented:
Yes, let's restrict the scope of the test to just what we care about for the purposes of this fix.

paavininanda commented:
Because of https://github.com/servo/servo/pull/20050, the idea of reverting `Revert the previous commit: git revert b2c1f89b932a72f9e0110c17adde33647e84c902` is creating merge conflicts.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #20051) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
I keep confusing myself between this PR and #20089. Am I right that #20089 is implementing the suggestions from https://github.com/servo/servo/pull/20069#issuecomment-367060272?

paavininanda commented:
> Am I right that #20089 is implementing the suggestions from #20069 (comment)?

Yes, I will try and do as you said 👍  And I think I should close this PR now ?

jdm commented:
Yes, I think so. Thanks!
